tests/scheduler_upgrade/scheduler_upgrade_test.go

```diff
@@ -0,0 +1,143 @@
+package tests
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/reporters"
+	. "github.com/onsi/gomega"
+	"github.com/portworx/torpedo/drivers/node"
+	"github.com/portworx/torpedo/drivers/scheduler"
+	. "github.com/portworx/torpedo/tests"
+	"github.com/sirupsen/logrus"
+
+	// https://github.com/kubernetes/client-go/issues/242
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+var (
+	upgradeTimeoutMins = 90 * time.Minute
+)
+
+func TestSchedulerUpgrade(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	var specReporters []Reporter
+	junitReporter := reporters.NewJUnitReporter("/testresults/junit_Upgrade.xml")
+	specReporters = append(specReporters, junitReporter)
+	RunSpecsWithDefaultAndCustomReporters(t, "Torpedo : SchedulerUpgrade", specReporters)
+}
+
+var _ = BeforeSuite(func() {
+	InitInstance()
+})
+
+var _ = Describe("{UpgradeScheduler}", func() {
+	var contexts []*scheduler.Context
+
+	It("upgrade scheduler and ensure everything is running fine", func() {
+		contexts = make([]*scheduler.Context, 0)
+
+		for i := 0; i < Inst().ScaleFactor; i++ {
+			contexts = append(contexts, ScheduleApplications(fmt.Sprintf("upgradescheduler-%d", i))...)
+		}
+
+		ValidateApplications(contexts)
+
+		upgradeHops := strings.Split(Inst().SchedUpgradeHops, ",")
+		Expect(len(upgradeHops)).NotTo(Equal(0), "No upgrade hops provided for UpgradeScheduler test")
+
+		intitialNodeCount, err := Inst().N.GetASGClusterSize()
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, schedVersion := range upgradeHops {
+			Step("start the upgrade of scheduler", func() {
+				err := Inst().N.SetClusterVersion(schedVersion, upgradeTimeoutMins)
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			Step(fmt.Sprintf("wait for %s minutes for auto recovery of storeage nodes",
+				Inst().AutoStorageNodeRecoveryTimeout.String()), func() {
+				time.Sleep(Inst().AutoStorageNodeRecoveryTimeout)
+			})
+
+			err = Inst().S.RefreshNodeRegistry()
+			Expect(err).NotTo(HaveOccurred())
+
+			err = Inst().V.RefreshDriverEndpoints()
+			Expect(err).NotTo(HaveOccurred())
+
+			Step(fmt.Sprintf("validate number of storage nodes after scheduler upgrade to [%s]",
+				schedVersion), func() {
+				ValidateClusterSize(intitialNodeCount)
+			})
+
+			Step("validate all apps after upgrade", func() {
+				for _, ctx := range contexts {
+					ValidateContext(ctx)
+				}
+			})
+		}
+	})
+})
+
+func ValidateClusterSize(count int64) {
+	// In multi-zone ASG cluster, node count is per zone
+	perZoneCount := count / 3
```

Comment:
It will fail if we run this test in non multi-zone cluster. If we assume that this test will be run in multi-zone cluster, then we have to add that restriction before we run this test.

Reply:
and also, we can run this test with MAX_STORAGE_NODES_PER_AZ=2 and the whole validation logic will fail here.

Reply:
@mborodin-px modified test to take list of zone into account so that test would work in single zone also. 
Consideration for MAX_STORAGE_NODES_PER_AZ has been taken care of here:
https://github.com/portworx/torpedo/pull/497/files#diff-8f145a99897a3709c5c8acaa2ad504cbR105